api/app/controllers/api/v1/offers_controller.rb

```diff
@@ -1,31 +1,396 @@
+    end
+
+    # can update DDAH status for offers that are accepted or pending
+    def can_ddah_update
+      check_offers_status(params[:offers], :status, ["Accepted", "Pending"])
+    end
+
+    # can send contracts for offers that are unsent or pending
+    def can_send_contract
+      check_offers_status(params[:offers], :status, ["Unsent", "Pending"])
+    end
+
+    # can nag for offers that are pending
+    def can_nag
+      check_offers_status(params[:offers], :status, ["Pending"])
+    end
+
+    # can print contracts for offers that are pending, accepted, or rejected
+    def can_print
+      check_offers_status(params[:offers], :status, ["Pending", "Accepted", "Rejected"])
+    end
+
+    def can_clear_hris_status
+      check_offers_status(params[:offers], :hr_status, [nil, "Processed", "Printed"])
+    end
+
+    def update_batch_offers_hours
+      params["_json"].each do |offer_json|
+        puts "looping"
+        puts offer_json
+        offer = Offer.find(offer_json[:offer_id])
+        puts offer
+        offer.update_attributes!(hours: offer_json[:hours])
+      end
+    end
+
+    def add_or_update
+      exceptions = []
+      offers = params[:offers]
+
+      offers.each do |offer|
+        position = Position.find_by(position: offer["course_id"])
+        applicant = Applicant.find_by(utorid: offer["utorid"])
+        if position && applicant
+          ident = {position_id: position[:id], applicant_id: applicant[:id]}
+          exists = "offer with position #{offer[:position]} for applicant #{offer[:utorid]} already exists"
+          data = {
+            position_id: position[:id],
+            applicant_id: applicant[:id],
+            hours: offer["hours"],
+            session: offer["session"],
+            year: offer["year"],
+          }
+          begin
+            insertion_helper(Offer, data, ident, exists)
+          rescue
+            exceptions.push("Error: Unknown error when creating Position '#{offer["course_id"]}' for '#{offer["utorid"]}' for '#{data[:hours]}' hours.")
+          end
+        else
+          if position
+            exceptions.push("Error: Applicant #{offer["utorid"]} is invalid.")
+          else
+            exceptions.push("Error: either Position #{offer["course_id"]} or Applicant #{offer["utorid"]} is invalid.")
+          end
+        end
+      end
+
+      if exceptions.length == offers.length
+        status = {success: false, errors: true, message: exceptions}
+      elsif exceptions.length > 0
+        status = {success: true, errors: true, message: exceptions}
+      else
+        status = {success: true, errors: false, message: ["Offers import was successful."]}
+      end
+
+      if status[:success]
+        render json: {errors: status[:errors], message: status[:message]}
+      else
+        render status: 404, json: {message: status[:message], errors: status[:errors]}
+      end
+    end
+
+    def update
```

Comment:
Only a specific set of Offer fields is mutable, in general Offers should not be updated.